### DISCUSSION TITLE
Virtual NTP server for private LAN

### DISCUSSION THREAD

**Discussion by rtek1000** [arendst/Tasmota#14293] [Q&A] [answered]
Considering that the main objective of the Tasmota project is to avoid the use of cloud servers located in unreliable countries, and therefore the effort to use your own local network:

It would be very helpful if a Tasmota firmware version was made available, which create a local synchronization service with NTP protocol.

In this way, the regular user, who does not have the understanding or conditions to compile in a custom way, could have a local server with reference to DS3231, even if it is manually configured. In this way all devices can be synchronized and timers can operate satisfactorily.

I've already done a project with the DS3231 to control a church tower clock and the stability of this IC is really good, even with a big temperature variation during the day.

I found some references:

https://tasmota.github.io/docs/DS3231/

https://forum.arduino.cc/t/looking-to-create-a-virtual-ntp-server-on-arduino-for-private-lan/133920

**Comment by ascillato** [comment, marked as answer]
Hi,

Local NTP server is already available in Tasmota. Please, check the docs at https://tasmota.github.io/docs/GPS-NTP-server/#commands

You will need to compile by yourself with NTP server support as explained in https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_60_GPS.ino

The NTP server works with GPS chip and also with DS3231 RTC chips.

More info at https://github.com/arendst/Tasmota/pull/7239

**Comment by rtek1000** [comment]
Hi, thanks for the information,

However, as far as I can understand, a server system based only on DS3231 is not yet available.

The options are:
- Use a Tasmota device with a GPS module as a time source
(https://tasmota.github.io/docs/GPS-NTP-server/)

or

- Use a DS3231 on each Tasmota device.
(https://tasmota.github.io/docs/DS3231/)

**Reply by sfromis** [reply]
One Tasmota device with a GPS as time source can act as a NTP server, if you activate that feature.

AFAICT, Tasmota with DS3231 cannot act as a NTP server, as this setup itself needs NTP server access to synchronize if it does not have a valid time.

**Reply by rtek1000** [reply]
Yes, it is an alternative that depends on an external network, the GPS system is a network of satellites, and like the Internet, it may not be available locally.

That's why I suggested a server with DS3231 only, so that the date and time can be adjusted locally manually, or with some sync button with a smartphone for example.

**Reply by pkkrusty** [reply]
I agree with this sentiment, would love to NTP server feature added to DS3231 code.

**Reply by pkkrusty** [reply]
I'm copying the NTP code from the GPS driver and will test it out in the next few days, but I think I've copied everything I need to, including using a bit in feature8 to turn it on and off. Sensor33 1 or 0. At least the firmware builds correctly, so my syntax is correct.

Uses the 1 second loop of the DS3231 driver, so not sure how that will interface with NTP client requests/timeouts and offsets, but if we can get within 1 second accuracy, that's probably good enough.

**Reply by pkkrusty** [reply]
Quick test indicates that it's working (another ESP8266 requested time from it and got it). Command to start and stop server works also. Current issues since I'm not actually good at programming:

Can't have `#use_gps` defined since `NtpServer timeServer(PortUdp);` in the declarations conflicts (already defined or something). But that will rarely be the case, since if you are building with `#use_gps`, you could use the pro-level NTP server. And also probably easily fixable, I just don't know the syntax.

Offset isn't calculated at all, I stuck 50ms in there because it sounded reasonable.

Submitted a PR so folks can try it out.

**Reply by pkkrusty** [reply]
PR merged #14318

**Comment by TrDA-hab** [comment]
The NTP server in the local network can be yours - a desktop or laptop, a smartphone or a router.
The operating system can be any - Windows or Unix/Linux, MacOS or others.
Maximum Tasmota supports 3 NTP servers. For example: two internal and one external NTP servers.

In my local network, this is a smart home server on Windows 10 and not big changes in the local policy.

**Reply by edegraaff** [reply]
Hi, short question:  i have internal my own ntp server, to make sure the tasmota devices sync in the right way, what settings should i set, and what settings should i remove?? In the past you added the timezone, and then program the tasmota with UTC as time driver. With summer and wintertime and a timeserver avalible i rather sync all devices with the same timesource, the NTP server on my router. Steps to approach this welcome :-)

**Reply by barbudor** [reply]
The commands 'ntpserver1' to 'ntpserver3' allows you to set alternative NTP servers that will be called in that order
It can be either ip address or name if your local server can be found name

**Reply by sfromis** [reply]
You cannot NTP sync to a time zone, as NTP is always UTC. And UTC is not a time zone.

You still need time zone config (`TimeSTD` and `TimeDST`) per device. Meaning that using the Timezone command is not really recommended, as Tasmota can handle the transitions when configured correctly.

**Reply by edegraaff** [reply]
Ok, but that would mean i need to reconfigure manually summer and wintertime? And thats what i am searchin for to bypass. The ntpserver1 to 3 are all set. From there what would be the best approach to have a " sit and forget"  config when it comes to summer and winter time?

**Reply by sfromis** [reply]
No, that is exactly what it does *not* mean, you can just setup the rules as already mentioned, for Tasmota to do it automatically. Here's a helper tool to generate the details for you, the "sit and forget" you ask for: https://tasmota-tz.netlify.app/

**Reply by edegraaff** [reply]
Ow, wow this is a great tool!
